### PR TITLE
[#116] SwagEdit does not detect some validation errors in some array …

### PR DESCRIPTION
**Pull request by ghillairet**
…schemas

See discussion in issue #116

### CODE REVIEW

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/validation/ValidatorTest.xtend [hunk 1]
```diff
@@ -525,4 +525,93 @@ class ValidatorTest {
...
+		        '200':
+		          description: OK
+		definitions:
+		  Foo:
+		    type: object
+		    properties:
+		      bar:
+		        type: string
+		    required:
+		      - baz
```
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/validation/ValidatorTest.xtend:606]: The error message here is `object has missing required properties %s`, where `%s` is the name of a property which is listed in the `required` list, but does not exist.  
To me something like `The referenced property %s does not exist`  or `property %s is not defined` would be more informative. 
@tedepstein , do you have suggestions for this case?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 2]
```diff
@@ -160,6 +177,74 @@ protected boolean hasArrayType(AbstractNode node) {
...
+    /**
+     * Validates an object type definition.
+     * 
+     * @param errors
+     * @param node
+     */
+    protected void checkObjectTypeDefinition(Set<SwaggerError> errors, AbstractNode node) {
+        if (node instanceof ObjectNode) {
+            JsonPointer ptr = node.getPointer();
+            if (ptr != null && ptr.toString().startsWith("/definitions")) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java]: Does it mean that this check won't be performed for inline types?

Comment (tfesenko) [reply com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java]: Yes, inline structures are not validated, e.g. for this model:

``` yaml
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /foo:
    get:
      summary: Read
      description: Provide details for the entire list (for collection resources) or an item (for object resources)
      responses:
        '200':
          description: OK
          schema:
            properties:
              bar:
                type: string
```

But how can we define if an element is a schema definition or not?...

**Review by tfesenko — COMMENTED** [review]
See my comment about checks that are only performed for schema definitions defined in `/definitions`, inline schemas are ignored.

**Comment by tfesenko** [comment]
Screenshot of new validation messages:
screen shot 2016-10-27 at 14 33 35
screen shot 2016-10-27 at 14 38 16
screen shot 2016-10-27 at 14 38 39

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/validation/ValidatorTest.xtend [hunk 3]
```diff
@@ -525,4 +525,93 @@ class ValidatorTest {
...
+		'''
+
+		document.set(content)
+		document.onChange()
+
+		val errors = validator.validate(document, null)		
+		assertEquals(1, errors.size())
+		assertEquals(String.format(Messages.error_required_properties, "baz"), errors.get(0).message)
+	}
```
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/validation/ValidatorTest.xtend:615]: Thank you for the tests!

**Comment by tfesenko** [comment]
@ghillairet this branch now has merge conflicts with the master

**Comment by ghillairet** [comment]
Inline schema are now supported.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 4]
```diff
@@ -188,9 +188,14 @@ protected boolean hasArrayType(AbstractNode node) {
...
         if (node instanceof ObjectNode) {
             JsonPointer ptr = node.getPointer();
-            if (ptr != null && ptr.toString().startsWith("/definitions")) {
-                checkMissingType(errors, node);
-                checkMissingRequiredProperties(errors, node);
+            if (ptr != null) {
+                if (ptr.toString().startsWith("/definitions")) {
+                    checkMissingType(errors, node);
+                    checkMissingRequiredProperties(errors, node);
+                } else if (ptr.toString().endsWith("/schema")) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:195]: I suggest extracting both checks (`ptrStr.startsWith("/definitions")` and `ptrStr.endsWith("/schema")`) into `isSchemaDefinition()`

**Review by tfesenko — APPROVED** [review]
Passed code review and QA.
Minor suggestion about code reusability for schema type checks.

**Pull request merged**